src/core/resolver/dns/c_ares/grpc_ares_wrapper.cc

```diff
@@ -1179,12 +1180,10 @@ static void grpc_cancel_ares_request_impl(grpc_ares_request* r) {
 
 void (*grpc_cancel_ares_request)(grpc_ares_request* r) =
     grpc_cancel_ares_request_impl;
+grpc_core::Mutex ares_setup_mu_;
 
-// ares_library_init and ares_library_cleanup are currently no-op except under
-// Windows. Calling them may cause race conditions when other parts of the
-// binary calls these functions concurrently.
-#ifdef GPR_WINDOWS
```

Comment:
Why removing this ifdef? I don't think we want to call these functions on non-Windows platforms.

Reply:
There are 2 parts answer to this question:

1. Consider this sequence of process in linux:
a. gRPC call `grpc_ares_init` (no-op)
b. curl call `ares_library_init` ([ares_initialized](https://github.com/c-ares/c-ares/blob/0e4b735a1060afa640b427c5c37df50f6014d991/src/lib/ares_library_init.c#L100C7-L100C23) incremented by one).
c. curl stop using ares, call `ares_library_cleanup`
d. gRPC process request, try to dns lookup using `grpc_dns_lookup_ares_continued` and eventually call [ares_init_options](https://github.com/grpc/grpc/blob/5638afba6eca657fe38c7566ac5a18004cec421f/src/core/resolver/dns/c_ares/grpc_ares_wrapper.cc#L532)

On our case, when (d) is happening, it introduces TSAN because ares is in the middle of cleanup while gRPC is still processing request. We can prevent it by calling `ares_library_init` on gRPC, so ares will know that there's another library that is still using ares (i.e. `ares_initialized` is still non-zero) and prevent it from cleanup.

2. After digging the issue more, I found out that we actually want to call this functions not on Windows, but on Android ([Code](https://github.com/c-ares/c-ares/blob/0e4b735a1060afa640b427c5c37df50f6014d991/src/lib/ares_library_init.c#L142) and [c-ares maintainer comment](https://github.com/c-ares/c-ares/issues/475#issuecomment-1166627950))

Reply:
I'm not sure we need to call this on Android, gRPC never called `ares_library_init_jvm()` nor `ares_library_init_android()`, so `ares_library_cleanup_android()` just looks like noop in this case. We haven't been doing that and haven't heard any issue yet.

`ares_init_options()` will call `ares_library_initialized()`: https://github.com/c-ares/c-ares/blob/main/src/lib/ares_init.c#L273 which seems to be noop on non-Windows platforms, so why TSAN complains? Can you post the TSAN error?

Reply:
Ack on no issue in the Android as of now.

`ares_library_initialized()` -> I don't think it's no-op. IIUC this line always return `ARES_SUCCESS` on non-Windows, that mean https://github.com/c-ares/c-ares/blob/main/src/lib/ares_init.c#L273 won't return early -> `ares_init_options` will continue as usual.

TSAN error can be seen on sponge2/306229f2-ef87-49bb-acb9-422486498e56

Reply:
Got it, so it's actually a race on accessing `ares_malloc`. With this change, wouldn't you still have the race on `ares_initialized`?

Reply:
Agree on `ares_initialized` is still not thread-safe, but I'm not sure the correct way to solve it without fixing c-ares, and the last attempt to fix it on c-ares is getting rejected on https://github.com/c-ares/c-ares/issues/475.

This PR only aims to prevent cleanup caused by `ares_library_cleanup` when gRPC is still using it (Which causes `ares_malloc` to be cleaned too early), and only fixing the thread-safety of `ares_initialized` won't fix this issue (At least that's what I tried).

Reply:
I think this change makes sense. You may need to add synchronization logic in your application around curl's init/cleanup and gRPC's init/shutdown or call them on the same thread to prevent this race.

@apolcyn WDYT?

Reply:
Friendly ping to this @apolcyn

Reply:
@apolcyn ping

Reply:
@apolcyn ping

Reply:
Apologies for the long review delay.

I'm not crazy about this change because, while fixing one thing, it looks like it potentially introduces a *new* bug: a data race on `ares_initialized` (if something outside grpc is also initializing/de-initializing c-ares concurrently with grpc).

Given that there is a potential data race i.e. undefined behavior at play, I'd opt to keep the current code as is since at least the existing bug is staying in the same place. I.e. this way we're less likely to cause a *new* hard-to-debug issue.

Here are a few other ideas that I think would be safer:

1) Have the application itself invoke `ares_library_init` once at the very top of main, before using `curl` or `grpc`.

2) Somehow prevent `curl` from invoking `ares_library_cleanup`. E.g. do not invoke invoke `curl_global_cleanup` or compile `curl` without `CARES_HAVE_ARES_LIBRARY_CLEANUP` (I am looking at [here](https://github.com/curl/curl/blob/0cececef0f70f9f349b185952bba6d387602bdd4/lib/asyn-ares.c#L146)

3) Maybe a bit hacky, but if we could patch c-ares to simply not [write](https://github.com/c-ares/c-ares/blob/a70637c67ba2657adb54b077f4f8434f00ecbf83/src/lib/ares_library_init.c#L155) to shared globals in `ares_library_cleanup` (i.e. leave `ares_malloc` and friends at the current state and only override them if subsequent initialization calls for it.

4) c-ares patch c-ares/c-ares#475 (I realize this has been rejected already, just listing for completeness)

5) Do the reverse patch of this - instead of patching c-ares to invoke `ares_library_init`/`ares_library_cleanup`, patch curl to *not* invoke this (if not running on windows).